test/TestProgramSettings.cc

```diff
@@ -214,4 +221,104 @@ TEST_F(ProgramSettingsTest, MultipleImagesFromPositional) {
     EXPECT_EQ(settings.files[0], casa_image.string().substr(1));
     EXPECT_EQ(settings.files[1], fits_image.string().substr(1));
     EXPECT_EQ(settings.files[2], hdf5_image.string().substr(1));
-}
\ No newline at end of file
+}
+
+TEST_F(ProgramSettingsTest, ExpectedValuesLongJSON) {
+    auto json_string = R"(
+    {
+        "verbosity": 6,
+        "no_log": true,
+        "no_http": true,
+        "no_browser": true,
+        "host": "helloworld",
+        "port": 1234,
+        "grpc_port": 5678,
+        "omp_threads": 10,
+        "top_level_folder": "/tmp",
+        "frontend_folder": "/var",
+        "exit_timeout": 10,
+        "initial_timeout": 11
+    })";
+    nlohmann::json j = nlohmann::json::parse(json_string);
+
+    carta::ProgramSettings settings;
+    settings.SetSettingsFromJSON(j);
+
+    EXPECT_EQ(settings.verbosity, 6);
+    EXPECT_EQ(settings.no_log, true);
+    EXPECT_EQ(settings.no_http, true);
+    EXPECT_EQ(settings.no_browser, true);
+    EXPECT_EQ(settings.host, "helloworld");
+    EXPECT_EQ(settings.port, 1234);
+    EXPECT_EQ(settings.grpc_port, 5678);
+    EXPECT_EQ(settings.omp_thread_count, 10);
+    EXPECT_EQ(settings.top_level_folder, "/tmp");
+    EXPECT_EQ(settings.frontend_folder, "/var");
+    EXPECT_EQ(settings.wait_time, 10);
+    EXPECT_EQ(settings.init_wait_time, 11);
+}
+
+TEST_F(ProgramSettingsTest, ValidateJSONFromFileWithGoodFields) {
+    const std::string input = "./data/settings-good-fields.json";
+    carta::ProgramSettings settings;
+    auto j = settings.JSONSettingsFromFile(input);
+    EXPECT_EQ(j.size(), 12);
+    EXPECT_EQ(j["verbosity"], 6);
+    EXPECT_EQ(j["port"], 1234);
+    EXPECT_EQ(j["grpc_port"], 5678);
+    EXPECT_EQ(j["omp_threads"], 10);
+    EXPECT_EQ(j["exit_timeout"], 10);
+    EXPECT_EQ(j["initial_timeout"], 11);
+    EXPECT_EQ(j["no_log"], true);
+    EXPECT_EQ(j["no_http"], true);
+    EXPECT_EQ(j["no_browser"], true);
+    EXPECT_EQ(j["host"], "helloworld");
+    EXPECT_EQ(j["top_level_folder"], "/tmp");
+    EXPECT_EQ(j["frontend_folder"], "/var");
+}
+
+TEST_F(ProgramSettingsTest, ValidateJSONFromFileWithBadFields) {
+    const std::string input = "./data/settings-bad-fields.json";
+    carta::ProgramSettings settings;
+    auto j = settings.JSONSettingsFromFile(input);
+    settings.SetSettingsFromJSON(j);
+    EXPECT_EQ(j.size(), 0);
+}
+
+TEST_F(ProgramSettingsTest, TestValuesFromGoodSettings) {
```

Comment:
I meant that in these tests we should parse a JSON string. We're still testing two functions at once here. The JSON string should contain the cleaned JSON with all bad values removed.

Reply:
Hi @confluence, I think that's what's done in `ValidateJSONFromFileWithGoodFields`:
``cpp
TEST_F(ProgramSettingsTest, ValidateJSONFromFileWithGoodFields) {
    const std::string input = "./data/settings-good-fields.json";
    carta::ProgramSettings settings;
    auto j = settings.JSONSettingsFromFile(input);
    EXPECT_EQ(j.size(), 12);
    EXPECT_EQ(j["verbosity"], 6);
    EXPECT_EQ(j["port"], 1234);
    EXPECT_EQ(j["grpc_port"], 5678);
    EXPECT_EQ(j["omp_threads"], 10);
    EXPECT_EQ(j["exit_timeout"], 10);
    EXPECT_EQ(j["initial_timeout"], 11);
    EXPECT_EQ(j["no_log"], true);
    EXPECT_EQ(j["no_http"], true);
    EXPECT_EQ(j["no_browser"], true);
    EXPECT_EQ(j["host"], "helloworld");
    EXPECT_EQ(j["top_level_folder"], "/tmp");
    EXPECT_EQ(j["frontend_folder"], "/var");
}
```

Reply:
>The message about the commandline options is the least relevant, but the message(s) about the locations of the config files could be useful (at debug level).

>Those messages should either be logged at debug level after the logger is initialised or not printed at all -- I think this is too verbose for info.

I agree with you. What I can do for the moment is to set them as debug. Still, the messages will never be printed since the logger defaults haven't changed at this point. We can later iterate on this behavior. What do you think? For the moment, I think that reducing this additional verbosity is important.

Reply:
I am tagging @veggiesaurus since he developed this part of the code, I think. 

What do you think if we initialize the logger immediately after parsing the first command-line settings if these include a higher level of verbosity? That would help to debug problems heere. If something else needs to be done in the current position of `InitLogger` in `Main.cpp`. We can reset call the function again. Am I missing something important there?

Reply:
1. `ValidateJSONFromFileWithGoodFields` and `ValidateJSONFromFileWithBadFields` test that `JSONSettingsFromFile` loads settings from a file correctly. `TestValuesFromGoodSettings` and `TestDefaultsFallbackFromBadSettings` should be testing **only** `SetSettingsFromJSON`. They should not be calling `JSONSettingsFromFile` at all. They should be constructing JSON objects directly from strings and passing those JSON objects to `SetSettingsFromJSON`. These objects should not contain any values with incorrect types, because those would have been removed by `JSONSettingsFromFile` in the actual code.
2. I would prefer not to leave in code that we know isn't working. If we aren't going to defer printing the config file location messages now, we should remove them entirely.

Reply:
Ok, I understand your point now. Thanks for explaining it further. I don't find that repeating the operation and reading from a file, then passing it to `SetSettingsFromJSON` is wrong. We know from the first test that the output from `JSONSettingsFromFile` is the expected (a good JSON and an empty JSON), then we want to check that `SetSettingsFromJSON` works properly. IMHO the first part is not being tested; at this point, it's only preparing the input for the test.

Reply:
Now that I'm thinking about this, I don't think we should print those warnings about bad keys immediately either. If the user wants to set the loglevel to > warning, we don't want to print warnings before the log is configured.

I would prefer to save any messages that arise during parsing of options, and print them later once the logger is initialised, than to move the initialisation of the logger earlier. I don't see how we could do the latter correctly -- we won't know what the verbosity is until we have finished parsing the options, so we shouldn't log messages before then.

If we're going to save the messages, we'll also need to save the loglevels -- so perhaps a list of level / message tuples. I'm not sure if spdlog has a generic function which takes a string level parameter; if not I guess we could save the actual functions by reference.

Reply:
That's a good point, indeed. I changed that code in the direction of your suggestion, but being a bit more empirical, with a vector storing warnings and debug messages that are flushed after the logger is initialized. I prefer this over a list of tuples that requires conditionals for deciding the print level.